spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/SkipperAppRegistryController.java

```diff
@@ -230,9 +249,58 @@ public void makeDefault(@PathVariable("type") ApplicationType type, @PathVariabl
 	@ResponseStatus(HttpStatus.OK)
 	public void unregister(@PathVariable("type") ApplicationType type, @PathVariable("name") String name,
 			@PathVariable("version") String version) {
+
+		// AppRegistration defaultApp = appRegistryService.getDefaultApp(name, type);
+		if (type != ApplicationType.task) { // && defaultApp != null && defaultApp.getVersion().equals(version)) {
+			String streamWithApp = findStreamContainingAppOf(type, name, version);
+			if (streamWithApp != null) {
+				throw new AppUsedInStreamException(String.format("The app [%s:%s:%s] you're trying to unregister is " +
+						"already in use on %s stream", name, type, version, streamWithApp));
+			}
+		}
+
 		appRegistryService.delete(name, type, version);
 	}
 
+	private String findStreamContainingAppOf(ApplicationType appType, String appName, String appVersion) {
+		Iterable<StreamDefinition> streams = streamDefinitionRepository.findAll();
+		for (StreamDefinition stream : streams) {
+			StreamDeployment streamDeployment = this.streamService.info(stream.getName());
+			for (StreamAppDefinition streamAppDefinition : stream.getAppDefinitions()) {
+				final String streamAppName = streamAppDefinition.getRegisteredAppName();
+				final ApplicationType streamAppType;
+				try {
+					streamAppType = DataFlowServerUtil.determineApplicationType(streamAppDefinition);
+				}
+				catch (CannotDetermineApplicationTypeException e) {
+					logger.warn("Can not determine ApplicationType for " + streamAppDefinition);
+					continue;
+				}
+				if (appType != streamAppType) {
+					break;
```

Comment:
I think this is wrong. Did some debugging with `time|log` when deleting `log`. It first sees time(source) which is different than log(sink) and breaks from for loop too early. Maybe this should be continue instead of break?